xdsl/backend/x86/lowering/convert_func_to_x86_func.py

```diff
@@ -0,0 +1,149 @@
+                "Cannot lower external functions (not implemented)"
+            )
+
+        for arg in first_block.args:
+            if isinstance(arg.type, builtin.ShapedType):
+                raise DiagnosticException(
+                    "Cannot lower shaped function parameters (not implemented)"
+                )
+            elif (
+                isinstance(arg.type, builtin.FixedBitwidthType)
+                and arg.type.bitwidth > 64
+            ):
+                raise DiagnosticException(
+                    "Cannot lower function parameters bigger than 64 bits (not implemented)"
+                )
+
+        num_inputs = len(op.function_type.inputs.data)
+        num_passing_args = num_inputs if num_inputs <= 6 else 6
+
+        insertion_point = InsertPoint.at_start(first_block)
+
+        # Get the 6 first parameters (if any) from general registers
+        reg_args_types = arg_passing_registers[0:num_passing_args]
+        actual_registers: list[SSAValue] = []
+        for register_type in reg_args_types:
+            get_reg_op = x86.ops.GetRegisterOp(register_type)
+            rewriter.insert_op(get_reg_op, insertion_point)
+            actual_registers.append(get_reg_op.result)
+
+        # Get the other parameters (if any) from the stack
+        if num_inputs > 6:
+            get_sp_op = x86.ops.GetRegisterOp(x86.register.RSP)
+            rewriter.insert_op(get_sp_op, insertion_point)
+            for i in range(num_inputs - 6):
+                get_reg_op = x86.ops.GetRegisterOp(x86.register.GeneralRegisterType(""))
+                mov_op = x86.RM_MovOp(
+                    r1=get_reg_op.result,
+                    r2=get_sp_op.result,
+                    offset=STACK_SLOT_SIZE_BYTES * (i + 1),
+                    result=x86.register.GeneralRegisterType(""),
+                )
+                actual_registers.append(mov_op.result)
+                rewriter.insert_op([get_reg_op, mov_op], insertion_point)
+
+        # Cast the registers to whatever type is needed
+        for arg, register in zip(first_block.args, actual_registers):
+            cast_op = builtin.UnrealizedConversionCastOp.get((register,), (arg.type,))
+            arg.replace_by(cast_op.results[0])
+            rewriter.insert_op(cast_op, insertion_point)
+
+        # Create the new function
+        new_region = rewriter.move_region_contents_to_new_regions(op.body)
+        block = new_region.blocks.first
+        assert isinstance(block, Block)
+        for a in block.args:
+            block.erase_arg(a)
+        new_func = x86_func.FuncOp(
+            op.sym_name.data,
+            new_region,
+            ([], []),
+            visibility=op.sym_visibility,
+        )
+
+        rewriter.replace_matched_op(new_func)
+
+
+class LowerReturnOp(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: func.ReturnOp, rewriter: PatternRewriter):
+        if len(op.arguments) == 0:
```

Comment:
this should be `if not op.arguments` instead, getting the length is potentially O(n), whereas truthiness is constant time and cheap